torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -494,6 +496,256 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   return producer;
 }
 
+// Create a TensorView before the original tensor. A common use case is to read
+// original tensor into shared memory or local registers.
+// Analogous to TVM Cache_Write
+TensorView* TensorView::cache_before() {
+  FusionGuard fg(this->fusion());
+
+  Expr* origin_expr = this->fusion()->origin(this);
+  TORCH_CHECK(
+      origin_expr != nullptr,
+      "Error adding cache_before ",
+      this,
+      " its origin is a nullptr and we restrict using cache_before on an input.");
+
+  // Create Producer Domain
+  auto dom = domain()->noReductions();
+  std::vector<IterDomain*> out_domain(dom.size(), nullptr);
+  for (size_t i = 0; i < dom.size(); ++i) {
+    if (dom[i]->isBroadcast()) {
+      out_domain[i] = new IterDomain(
+          new Int(0), new Int(1), ParallelType::Serial, false, false, true);
+    } else {
+      out_domain[i] = new IterDomain(dom[i]->start(), dom[i]->extent());
```

Comment:
No reason not to copy over all information that isn't broadcast and reduction. I think just parallel type is left, but check.